test_utils/utils/create_partition.py

```diff
@@ -0,0 +1,246 @@
+		):
+			return
+
+		custom_field = frappe.get_doc(
+			{
+				"doctype": "Custom Field",
+				"dt": child_doctype,
+				"fieldname": partition_docfield.fieldname,
+				"fieldtype": partition_docfield.fieldtype,
+				"label": partition_docfield.label,
+				"options": partition_docfield.options,
+				"default": partition_docfield.default,
+				"read_only": 1,
+				"hidden": 1,
+			}
+		)
+		try:
+			custom_field.insert()
+		except Exception:
+			continue
+
+
+def populate_partition_fields(doc, event):
+	"""
+	doc_events = {
+	        "*": {
+	                "before_insert": "yourapp.utils.populate_partition_fields",
+	                "before_save": "yourapp.utils.populate_partition_fields",
+	        }
+	}
+	"""
+	partition_doctypes = frappe.get_hooks("partition_doctypes")
+
+	if doc.doctype not in partition_doctypes.keys():
+		return
+
+	partition_field = partition_doctypes[doc.doctype]["field"][0]
+
+	for child_doctype_fieldname in [
+		(df.options, df.fieldname) for df in frappe.get_meta(doc.doctype).get_table_fields()
+	]:
+		child_doctype = child_doctype_fieldname[0]
+		child_fieldname = child_doctype_fieldname[1]
+
+		if not frappe.get_meta(child_doctype)._fields.get(partition_field):
+			add_custom_field(doc.doctype, partition_field)
+
+		for row in getattr(doc, child_fieldname):
+			setattr(row, partition_field, doc.get(partition_field))
+
+
+def primary_key_exists(table_name):
+	try:
+		result = frappe.db.sql(
+			f"""
+		SELECT COUNT(*)
+		FROM information_schema.TABLE_CONSTRAINTS
+		WHERE TABLE_NAME = '{table_name}'
+		AND CONSTRAINT_TYPE = 'PRIMARY KEY';
+		"""
+		)
+		return result[0][0] > 0
+	except Exception as e:
+		print(f"Error checking primary key existence: {e}")
+		return False
+
+
+def modify_primary_key(table_name, partition_field):
+	try:
+		if primary_key_exists(table_name):
+			drop_pk_sql = f"""
+			ALTER TABLE `{table_name}`
+			DROP PRIMARY KEY;
+			"""
+			frappe.db.sql(drop_pk_sql)
+
+			unique_indexes = frappe.db.sql(
+				f"SHOW INDEXES FROM `{table_name}` WHERE Non_unique = 0", as_dict=True
+			)
+
+			for index in unique_indexes:
+				index_name = index["Key_name"]
+				columns = index["Column_name"]
+				frappe.db.sql(f"ALTER TABLE `{table_name}` DROP INDEX `{index_name}`;")
+				if partition_field not in columns:
+					columns = f"{columns}, `{partition_field}`"
+				frappe.db.sql(
+					f"ALTER TABLE `{table_name}` ADD UNIQUE INDEX `{index_name}` ({columns});"
+				)
+
+			pk_columns = f"name, {partition_field}"
+			add_pk_sql = f"""
+			ALTER TABLE `{table_name}`
+			ADD PRIMARY KEY ({pk_columns});
+			"""
+			frappe.db.sql(add_pk_sql)
+			frappe.db.commit()
+			print(f"Primary key modified in table {table_name} to include columns: {pk_columns}")
+	except Exception as e:
+		print(f"Error modifying primary key in table {table_name}: {e}")
+
+
+def get_partition_doctypes_extended():
+	partition_doctypes = frappe.get_hooks("partition_doctypes")
```

Comment:
@agritheory The current behavior is to manage automatically all the child doctypes for each doctype declared en `partition_doctypes`. We have other options if this isn't suited to our needs.

- Add an option in the `partition_doctypes` like `no_include_child_doctypes` or `include_child_doctypes`.
- Add an option in the `partition_doctypes` to explicitly declare which child doctypes we want to get partitioned.

Reply:
I think automatic might be better.